sklearn/linear_model/tests/test_coordinate_descent.py

```diff
@@ -111,7 +113,10 @@ def test_lasso_zero():
     # Check that the lasso can handle zero data without crashing
     X = [[0], [0], [0]]
     y = [0, 0, 0]
-    clf = Lasso(alpha=0.1).fit(X, y)
+    # _cd_fast.pyx tests for gap < tol, but here we get 0.0 < 0.0
```

Comment:
Actually I do not understand why it would be the case in this test. By default the tolerance is 1e-4:

```python
>>> from sklearn.linear_model import Lasso
>>> Lasso().tol
0.0001
```

This test does not override the default. The same is true for `MultiTaskLasso`.

Reply:
Ah ok I understand, this is because we rescale the tolerance by a data dependent quantity which is 0 in this case.

Reply:
That is indeed sub-optimal - `y == 0` are valid options. What about the following change to keep `tol` as is if it would become `0` (similar code for all the different coordinate descent function instances):
```
--- a/sklearn/linear_model/_cd_fast.pyx
+++ b/sklearn/linear_model/_cd_fast.pyx
@@ -672,6 +672,7 @@ def enet_coordinate_descent_multi_task(
     cdef floating W_ii_abs_max
     cdef floating gap = tol + 1.0
     cdef floating d_w_tol = tol
+    cdef floating temp_tol = tol
     cdef floating R_norm
     cdef floating w_norm
     cdef floating ry_sum
@@ -704,7 +705,9 @@ def enet_coordinate_descent_multi_task(
                           &R[0, jj], 1)
 
         # tol = tol * linalg.norm(Y, ord='fro') ** 2
-        tol = tol * _nrm2(n_samples * n_tasks, Y_ptr, 1) ** 2
+        temp_tol = tol * _nrm2(n_samples * n_tasks, Y_ptr, 1) ** 2
+        if temp_tol > 0:
+            tol = temp_tol
 
         for n_iter in range(max_iter):
             w_max = 0.0
@@ -819,7 +822,7 @@ def enet_coordinate_descent_multi_task(
                 gap += l1_reg * l21_norm - const * ry_sum + \
                      0.5 * l2_reg * (1 + const ** 2) * (w_norm ** 2)
 
-                if gap <= tol:
+                if gap < tol:
                     # return if we reached desired tolerance
                     break
         else:
```
Then we could also return to the previous (`gap < tol`).